### ISSUE TITLE
Referencing list items

### ISSUE THREAD

**Issue by hseg** [lierdakil/pandoc-crossref#124]
It would be nice to be able to emulate LaTeX's
```latex
TFAE:
\begin{enumerate}
\item\label{def1} foo
\item\label{def2} bar
\item\label{def3} baz
\end{enumerate}

Proof: \ref{def1} implies \ref{def2}:...
```
The absence of such a feature is sorely missing for me.
Suggested syntax:
```markdown
TFAE:
(#it:def1). foo
(#it:def2). bar
(#it:def3). baz

Proof: -@it:def1 implies -@it:def2:...
```
For the moment, this can hackily be worked around using
```markdown
TFAE:
- $$\text{foo}$$ {#eq:def1}
- $$\text{bar}$$ {#eq:def2}
- $$\text{baz}$$ {#eq:def3}

Proof: @eq:def1 implies @eq:def2:...
```
but this is clearly subpar.

**Comment by lierdakil** [comment]
We have to work mostly within the constraints of Pandoc syntax, so suggested syntax won't work.

Besides, list item cross-references differ from all other types in that numeration is not document-wide. I.e. you can have different items with identical numbers, and that can (and will) be a problem. If you want document-wide numeration, [example lists](http://pandoc.org/MANUAL.html#numbered-example-lists) already do what you want.

**Comment by hseg** [comment]
Sorry for late response.
Considering this issue, the only cases where we would want to refer to a list item is when
it is contained in another, globally-unique, enumeration
(think clauses of a definition, equivalent characterizations, etc.)
The issue is that unlike LaTeX, Pandoc doesn't support defining list types which have different counters.
Otherwise, we could say something along the lines of:

```
(#dfn:ring) A _ring_ is a triple $(R,+,*)$ satisfying:
    (#cl:addgp) $+$ is an abelian group
    (#cl:multmon) $*$ is a monoid
    (#cl:distrib) $*$ distributes over $+$
(#prop:elem-ring) Some elementary properties of rings:
    (#cl:mult-neg1-is-neg)    $-a=(-1)*a$

        Proof: Using @cl:multmon and @cl:distrib, $a+(-1)*a=1*a+(-1)*a=(1-1)*a=0*a$
```
with e.g. `@cl:multmon` expanding to the indeces of `#dfn:ring` and `#cl:multmon`, separated by a `.`

Basically, the idea is that "clause-type" lists would pull in the index of the list containing them,
and would be brequired to be contained in a non-clause-type list.
Add to that the ability to separate enumeration of e.g. definitions, propositions, etc.
and you basically subsume LaTeX's enumeration abilities.
(we might want to add the capacity for identifying some counters, though)

Of course, this might be beyond the scope of what we want.

And yes, considering what you've said regarding having to work within Pandoc syntax,
this will probably require bikeshedding here.
One possibility is to subsume example list syntax and write
```
(@dfn:ring) ...
    (@cl:addgp)
...
    Proof: Using (@cl:addgp)...
```

What do you think?

**Comment by lierdakil** [comment]
> "clause-type" lists would pull in the index of the list containing them,
and would be required to be contained in a non-clause-type list.

This looks somewhat overcomplicated tbh. This boils down to arbitrary nesting of arbitrary counters... which would be cool, but that seems somewhat hard to design, implement and use. Unless there's some sort of universal rule one could apply, this will quickly devolve into a mess of ten thousand special cases...

> we might want to add the capacity for identifying some counters, though

I've thought about it before. Didn't reach any sort of a conclusion, but it will have to be through metadata and prefixing, I reckon. For example:

```yaml
prefixes:
  dfn:
    ref: ["dfn.", "dfns."]
    name: "Definition"
  cl:
    ref: ["cl.", "cls."]
    scope: ["dfn"]
```

Pretty hazy on scoping rules though.

> One possibility is to subsume example list syntax

Sadly, example list syntax is kinda out of reach too, since it's substituted at the time of parsing... so no real way to get at that from a filter. Frankly our best bet would be arbitrary div/span elements (and raw LaTeX inclusions if we really have to). For example (using new span syntax):

```markdown
<div id="dfn:ring">
A _ring_ is a triple $(R,+,*)$ satisfying:

#. [$+$ is an abelian group]{#cl:addgp}
#. [$*$ is a monoid]{#cl:multmon}
#. [$*$ distributes over $+$]{#cl:distrib}
</div>
```

This seems somewhat awkward though.

**Comment by lierdakil** [comment]
I'll close this in favor of #84.
